gitlab/utils.py

```diff
@@ -56,25 +56,38 @@ def copy_dict(dest: Dict[str, Any], src: Dict[str, Any]) -> None:
             dest[k] = v
 
 
-def _url_encode(id: str) -> str:
-    """Encode/quote the characters in the string so that they can be used in a path.
+class EncodedId(str):
+    """A custom `str` class that will return the URL-encoded value of the string.
 
-    Reference to documentation on why this is necessary.
-
-    https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding
-
-    If using namespaced API requests, make sure that the NAMESPACE/PROJECT_PATH is
-    URL-encoded. For example, / is represented by %2F
+      * Using it recursively will only url-encode the value once.
+      * Can accept either `str` or `int` as input value.
+      * Can be used in an f-string and output the URL-encoded string.
 
-    https://docs.gitlab.com/ee/api/index.html#path-parameters
+    Reference to documentation on why this is necessary.
 
-    Path parameters that are required to be URL-encoded must be followed. If not, it
-    doesn’t match an API endpoint and responds with a 404. If there’s something in front
-    of the API (for example, Apache), ensure that it doesn’t decode the URL-encoded path
-    parameters.
+    See::
 
+        https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding
+        https://docs.gitlab.com/ee/api/index.html#path-parameters
     """
-    return urllib.parse.quote(id, safe="")
+
+    # `original_str` will contain the original string value that was used to create the
+    # first instance of EncodedId. We will use this original value to generate the
+    # URL-encoded value each time.
+    original_str: str
+
```

Comment:
I think we can now remove this as it's not used anywhere:

```suggestion
```

Reply:
Done